File: src/lib/db/writeDiscovery.ts
/**
 * writeDiscovery — write path for discovery agent runs.
 *
 * Writes to:
 *   1. Firestore businesses/{slug} — creates/updates the business document with current identity
 *   2. BigQuery hephae.discoveries — permanent append-only record of every discovery run
 *
 * RULE: raw_data must never contain menuScreenshotBase64 or any binary blob.
 *       Pass menuImageUrl (GCS URL) instead.
 */

import { Timestamp, FieldValue } from 'firebase-admin/firestore';
import { db } from '@/lib/firebase';
import { bq, DATASET } from './bigquery';
import { EnrichedProfile } from '@/agents/types';
import { AgentVersions } from '@/agents/config';

/**
 * Strips binary blobs from an EnrichedProfile before any database write.
 * menuScreenshotBase64 must be uploaded to GCS and replaced with a URL before calling this.
 */
export function stripBlobs(profile: EnrichedProfile): Omit<EnrichedProfile, 'menuScreenshotBase64'> {
    const { menuScreenshotBase64: _stripped, ...safe } = profile;
    return safe;
}

/**
 * Parse zip code from an address string as a best-effort fallback.
 * Prefer passing zipCode explicitly when available.
 */
function parseZipCode(address?: string): string | null {
    if (!address) return null;
    const match = address.match(/\b(\d{5})(?:-\d{4})?\b/);
    return match ? match[1] : null;
}

export interface WriteDiscoveryOptions {
    profile: EnrichedProfile;
    menuImageUrl?: string;       // GCS URL to menu screenshot — replaces base64
    zipCode?: string;            // explicit zip; falls back to parsing from address
    triggeredBy: 'user' | 'weekly_job' | 'api_v1';
}

export async function writeDiscovery(opts: WriteDiscoveryOptions): Promise<void> {
    const { profile, menuImageUrl, triggeredBy } = opts;
    const runAt = new Date();
    const runId = `discovery-${runAt.getTime()}`;

    const zipCode = opts.zipCode ?? parseZipCode(profile.address) ?? undefined;

    // Never write base64 blobs to Firestore or BQ
    const safe = stripBlobs(profile);

    // --- 1. Firestore upsert ---
    try {
        await db.doc(`businesses/${profile.name.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/^-|-$/g, '')}`).set({
            name: profile.name,
            address: profile.address ?? null,
            officialUrl: profile.officialUrl,
            coordinates: profile.coordinates ?? null,
            ...(zipCode ? { zipCode } : {}),
            updatedAt: Timestamp.fromDate(runAt),
            createdAt: FieldValue.serverTimestamp(),   // ignored if doc already exists via merge
            identity: {
                phone: profile.phone ?? null,
                email: profile.email ?? null,
                hours: profile.hours ?? null,
                googleMapsUrl: profile.googleMapsUrl ?? null,
                socialLinks: profile.socialLinks ?? {},
                logoUrl: profile.logoUrl ?? null,
                favicon: profile.favicon ?? null,
                primaryColor: profile.primaryColor ?? null,
                secondaryColor: profile.secondaryColor ?? null,
                persona: profile.persona ?? null,
                menuImageUrl: menuImageUrl ?? null,
                competitors: profile.competitors ?? [],
            },
        }, { merge: true });
    } catch (err) {
        console.error(`[DB] Firestore writeDiscovery failed for ${profile.name}:`, err);
    }

    // --- 2. BigQuery append ---
    const row = {
        run_id: runId,
        business_slug: profile.name.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/^-|-$/g, ''),
        business_name: profile.name,
        official_url: profile.officialUrl,
        address: profile.address ?? null,
        city: null,   // future: parse from address
        state: null,  // future: parse from address
        zip_code: zipCode ?? null,
        lat: profile.coordinates?.lat ?? null,
        lng: profile.coordinates?.lng ?? null,
        agent_name: 'discovery_orchestrator',
        agent_version: AgentVersions.MENU_DISCOVERY,  // orchestrator version tracks the pipeline
        run_at: runAt.toISOString(),
        triggered_by: triggeredBy,
        raw_data: JSON.stringify({
            ...safe,
            menuImageUrl: menuImageUrl ?? null,
        }),
    };

    bq.dataset(DATASET).table('discoveries').insert([row], { skipInvalidRows: false })
        .catch(err => {
            console.error(`[DB] BQ discoveries write failed for ${runId}:`, err?.errors ?? err);
        });
}

File: backend/lib/db/write_discovery.py
"""
writeDiscovery — write path for discovery agent runs.

Writes to:
  1. Firestore businesses/{slug} — creates/updates the business document with current identity
  2. BigQuery hephae.discoveries — permanent append-only record of every discovery run

RULE: raw_data must never contain menuScreenshotBase64 or any binary blob.
      Pass menuImageUrl (GCS URL) instead.
"""

from __future__ import annotations

import asyncio
import json
import logging
import re
from datetime import datetime, timezone
from typing import Any, Optional

from google.cloud.firestore_v1 import SERVER_TIMESTAMP

from backend.config import AgentVersions
from backend.lib.report_storage import generate_slug

logger = logging.getLogger(__name__)


def strip_blobs(profile: dict[str, Any]) -> dict[str, Any]:
    """
    Strip binary blobs from an EnrichedProfile dict before any database write.
    menuScreenshotBase64 must never reach the DB.
    """
    safe = {k: v for k, v in profile.items() if k != "menuScreenshotBase64"}
    return safe


def _parse_zip_code(address: Optional[str]) -> Optional[str]:
    """Parse zip code from an address string as a best-effort fallback."""
    if not address:
        return None
    match = re.search(r"\b(\d{5})(?:-\d{4})?\b", address)
    return match.group(1) if match else None


async def write_discovery(
    profile: dict[str, Any],
    triggered_by: str = "user",
    zip_code: Optional[str] = None,
) -> None:
    """Write discovery result to Firestore + BigQuery."""
    from backend.lib.firebase import db
    from backend.lib.bigquery import bq_insert

    run_at = datetime.now(timezone.utc)
    run_id = f"discovery-{int(run_at.timestamp() * 1000)}"

    resolved_zip = zip_code or _parse_zip_code(profile.get("address"))
    slug = generate_slug(profile.get("name", "unknown"))

    # Never write base64 blobs to Firestore or BQ
    safe = strip_blobs(profile)

    # --- 1. Firestore upsert ---
    try:
        # Enriched fields live at the top level for direct queries AND
        # inside identity for backwards compatibility
        enriched_fields = {
            "phone": profile.get("phone"),
            "email": profile.get("email"),
            "hours": profile.get("hours"),
            "googleMapsUrl": profile.get("googleMapsUrl"),
            "socialLinks": profile.get("socialLinks", {}),
            "logoUrl": profile.get("logoUrl"),
            "favicon": profile.get("favicon"),
            "primaryColor": profile.get("primaryColor"),
            "secondaryColor": profile.get("secondaryColor"),
            "persona": profile.get("persona"),
            "menuUrl": profile.get("menuUrl"),
            "menuScreenshotUrl": profile.get("menuScreenshotUrl"),
            "menuHtmlUrl": profile.get("menuHtmlUrl"),
            "competitors": profile.get("competitors", []),
            "socialProfileMetrics": profile.get("socialProfileMetrics"),
            "news": profile.get("news"),
            "aiOverview": profile.get("aiOverview"),
            "validationReport": profile.get("validationReport"),
        }

        doc_data: dict[str, Any] = {
            "name": profile.get("name"),
            "address": profile.get("address"),
            "officialUrl": profile.get("officialUrl", ""),
            "coordinates": profile.get("coordinates"),
            "updatedAt": run_at,
            "createdAt": SERVER_TIMESTAMP,
            # Top-level enriched fields
            **enriched_fields,
            # Also keep identity sub-object for backwards compatibility
            "identity": enriched_fields,
        }
        if resolved_zip:
            doc_data["zipCode"] = resolved_zip

        db.document(f"businesses/{slug}").set(doc_data, merge=True)
    except Exception as err:
        logger.error(f"[DB] Firestore writeDiscovery failed for {profile.get('name')}: {err}")

    # --- 2. BigQuery append ---
    coords = profile.get("coordinates") or {}
    row = {
        "run_id": run_id,
        "business_slug": slug,
        "business_name": profile.get("name"),
        "official_url": profile.get("officialUrl", ""),
        "address": profile.get("address"),
        "city": None,
        "state": None,
        "zip_code": resolved_zip,
        "lat": coords.get("lat"),
        "lng": coords.get("lng"),
        "agent_name": "discovery_orchestrator",
        "agent_version": AgentVersions.MENU_DISCOVERY,
        "run_at": run_at,
        "triggered_by": triggered_by,
        "raw_data": json.dumps(safe),
    }

    asyncio.get_event_loop().run_in_executor(
        None,
        lambda: _bq_insert_sync(bq_insert, "discoveries", row, run_id),
    )


def _bq_insert_sync(bq_insert_fn, table: str, row: dict, run_id: str) -> None:
    """Synchronous wrapper for fire-and-forget BQ insert."""
    import asyncio

    try:
        asyncio.run(bq_insert_fn(table, row))
    except Exception as err:
        logger.error(f"[DB] BQ discoveries write failed for {run_id}: {err}")
